lms/resources/_js_config.py

```diff
@@ -131,6 +135,13 @@ def enable_canvas_oauth2_redirect_error_mode(
         )
 
     def enable_lti_launch_mode(self):
+        """
+        Put the JavaScript code into "LTI launch" mode.
+
+        This mode launches an assignment.
+
+        :raise ConsumerKeyError: if request.lti_user.consumer_key isn't in the DB
+        """
```

Comment:
I agree, but with the way the code is I found it quite helpful / necessary to trace all the places that could raise `ConsumerKeyError` before and that can raise it now in order to verify that I hadn't made any unintentional behaviour changes.

I think we should consider lots of stacked `:raise:`'s a code smell. I think there's two mitigations. One is that the code shouldn't have too many layers. In this case `ApplicationInstanceService` raises `ConsumerKeyError` and `JSConfig` calls `ApplicationInstanceService` so `JSConfig` also raises `ConsumerKeyError`. The only other layer (that calls `JSConfig`) is the view, and the particular views that call it would never trigger the `ConsumerKeyError` since authentication means that the view can't be called without a valid `ApplicationInstance` (or in the case of the views that _can_ be called without an `ApplicationInstance`, they don't call any of the `JSConfig` methods that might raise). So we come to a stop quite quickly, in tracing all the things that might raise this `ConsumerKeyError`. Still we could have one layer here instead of two.

The other thing I think is maybe a design issue with `JSConfig` where a context resource is wrapping a service, and has lots of methods that might raise. There may be a better design there, such as making sure that a `JSConfig` can't even be constructed in the first place if there isn't a valid `ApplicationInstance`. Or just having views call a service directly rather than using a resource (so make a `JSConfigService`).

I think Pyramid's "context resource" pattern is sometimes useful, but only in quite specific ways. For example the idea of an annotation update API view, where the `models.Annotation` to be updated (or an `AnnotationResource` that wraps `models.Annotation`) gets passed in to the view as the `context` argument and the view can then pass it to services. That seems quite useful. But I think with `LTILaunchResource` and `JSConfig` the LMS app is doing things in resources that really belong in services. Really, a lot of code makes more sense in services than anywhere else, and other layers like resources shouldn't be over-used.